[min-cut partitioner] Disable a heuristic if graph has recomputable ops

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103601
* __->__ #103635


Removing this heuristic leads to major memory compression and speedup bump for activation-checkpointed models. Here is the data
![image](https://github.com/pytorch/pytorch/assets/13822661/64a491ab-173d-435a-b858-61b847fbb08b)

anijain2305 committed:
[min-cut partitioner] Disable a heuristic if graph has recomputable ops

[ghstack-poisoned]

anijain2305 committed:
Update on "[min-cut partitioner] Disable a heuristic if graph has recomputable ops"




[ghstack-poisoned]

anijain2305 committed:
Update on "[min-cut partitioner] Disable a heuristic if graph has recomputable ops"




[ghstack-poisoned]

anijain2305 committed:
Update on "[min-cut partitioner] Disable a heuristic if graph has recomputable ops"



Removing this heuristic leads to major memory compression and speedup bump for activation-checkpointed models. Here is the data
![image](https://github.com/pytorch/pytorch/assets/13822661/64a491ab-173d-435a-b858-61b847fbb08b)



[ghstack-poisoned]

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -783,8 +783,12 @@ def ban_recomputation(node):
+            # TODO: Investigate the interaction with compiler assisted
+            # activation checkpointing. Removing the heuristic improves both
+            # memory footprint and speedup.
+            if not graph_has_recomputable_ops:
```
Can we not just do the same thing where we apply it to each node individually and not the entire graph at once?

anijain2305 replied:
I don't think we can apply per node here. This line is reached only for those nodes that do not have recompute tags (see L760 below).

https://github.com/pytorch/pytorch/blob/2adfd1315aaf45a356a8570f89054a6736aebb1f/torch/_functorch/partitioners.py#L760-L787

This PR is saying that for nodes sitting outside of activation checkpointing, lets disable the heuristic for them, and let min-cut partitioner find the cut. I looked at couple of models, and the saved tensor size reduction is quite large (like 30% - 50%).

Chillee replied:
Ok, I don't understand why this helps (but tbh, i also don't have a great understanding why the current heuristic helps). Would be nice to resolve those at some point.

anijain2305 replied:
Yes, I tried to understand it as well. But wasn't able to see this on a small model. On large models, it gets difficult to reason why this works. 

For one model, it was pretty clear though. There was some `full` operator which was producing a large tensor, and we were stashing some intermediate tensor far away from that `full` op. With heuristic disabled, it went all the way to the full and saved memory footprint.

Chillee reviewed (approved):
Will approve for now, but I would like a resolution to the comment.

anijain2305 committed:
Update on "[min-cut partitioner] Disable a heuristic if graph has recomputable ops"



Removing this heuristic leads to major memory compression and speedup bump for activation-checkpointed models. Here is the data
![image](https://github.com/pytorch/pytorch/assets/13822661/64a491ab-173d-435a-b858-61b847fbb08b)



[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge -f "rocm tests passed last time, expecting them to pass"
